shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/transaction/TransactionXAHandler.java

```diff
@@ -0,0 +1,105 @@
+ */
+@RequiredArgsConstructor
+public final class TransactionXAHandler implements TextProtocolBackendHandler {
+
+    private final XAStatement tclStatement;
+
+    private final ConnectionSession connectionSession;
+
+    private final SchemaAssignedDatabaseBackendHandler backendHandler;
+
+    public TransactionXAHandler(final SQLStatementContext<? extends TCLStatement> sqlStatementContext, final String sql, final ConnectionSession connectionSession) {
+        this.tclStatement = (XAStatement) sqlStatementContext.getSqlStatement();
+        this.connectionSession = connectionSession;
+        this.backendHandler = new SchemaAssignedDatabaseBackendHandler(sqlStatementContext, sql, connectionSession);
+    }
+
+    @Override
+    public boolean next() throws SQLException {
+        return this.tclStatement.getOp().equals("RECOVER") && this.backendHandler.next();
+    }
+
+    @Override
+    public Collection<Object> getRowData() throws SQLException {
+        return this.tclStatement.getOp().equals("RECOVER") ? this.backendHandler.getRowData() : Collections.emptyList();
+    }
+
+    @Override
+    public ResponseHeader execute() throws SQLException {
+        switch (tclStatement.getOp()) {
+            case "START":
+            case "BEGIN":
+                /**
+                 * we have to let session occupy the thread when doing xa transaction.
+                 * according to https://dev.mysql.com/doc/refman/5.7/en/xa-states.html XA and local transactions are mutually exclusive
+                 */
+                if (connectionSession.getTransactionStatus().isInTransaction()) {
+                    if (TransactionType.MANUALXA != connectionSession.getTransactionStatus().getTransactionType()) {
+                        BackendTransactionManager backendTransactionManager = new BackendTransactionManager((JDBCBackendConnection) connectionSession.getBackendConnection());
+                        backendTransactionManager.commit();
+                    } else {
+                        throw new SQLException("already in XA transaction");
+                    }
+                }
+                TransactionHolder.setInTransaction();
+                connectionSession.setAutoCommit(false);
+                connectionSession.getTransactionStatus().setManualXA(true);
+                JDBCBackendConnection connection = (JDBCBackendConnection) connectionSession.getBackendConnection();
+                connection.closeDatabaseCommunicationEngines(true);
+                connection.closeConnections(false);
+                return backendHandler.execute();
+            case "END":
+            case "PREPARE":
+            case "RECOVER":
+                return backendHandler.execute();
+            case "COMMIT":
+            case "ROLLBACK":
+                try {
+                    return backendHandler.execute();
+                } finally {
+                    connectionSession.getTransactionStatus().setManualXA(false);
```

Comment:
It also need to ```setInTransaction(false)```

Reply:
yes, override `isInTransaction()` in `TransactionStatus` to avoid set it.